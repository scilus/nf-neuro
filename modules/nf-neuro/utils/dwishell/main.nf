process UTILS_DWISHELL {
    tag "${meta.id}"
    label 'process_single'

    container "scilus/scilpy:2.2.0_cpu"

    input:
    tuple val(meta), path(dwi), path(bval), path(bvec)

    output:
    tuple val(meta), path("*__dwi_sh_shells.nii.gz"), emit: dwi
    tuple val(meta), path("*__bval_sh_shells"), emit: bval
    tuple val(meta), path("*__bvec_sh_shells"), emit: bvec
    tuple val(meta), path("*__out_indices"), emit: out_indices, optional: true
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"

    def out_indices = task.ext.out_indices ? "--out_indices ${prefix}__out_indices" : ''
    def block_size = task.ext.block_size ? "--block_size ${task.ext.block_size}" : ''
    def tolerance = task.ext.tolerance ? "--tolerance ${task.ext.tolerance}" : ''
    def max_shell_bvalue = task.ext.max_shell_bvalue ?: 1500
    def b0_thr_extract_b0 = task.ext.b0_thr_extract_b0 ?: 10
    def shell_to_fit = task.ext.shell_to_fit ?: "\$(cut -d ' ' --output-delimiter=\$'\\n' -f 1- ${bval} | awk -F' ' '{v=int(\$1)}{if(v<=${max_shell_bvalue}|| v<=${b0_thr_extract_b0})print v}' | sort | uniq)"
    def force = task.ext.force ? "-f" : ''

    """
    scil_dwi_extract_shell ${dwi} ${bval} ${bvec} ${shell_to_fit} \
        ${prefix}__dwi_sh_shells.nii.gz \
        ${prefix}__bval_sh_shells \
        ${prefix}__bvec_sh_shells \
        ${out_indices} \
        ${block_size} \
        ${tolerance} \
        ${force}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(uv pip -q -n list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    scil_dwi_extract_shell -h
    touch ${prefix}__dwi_sh_shells.nii.gz
    touch ${prefix}__bval_sh_shells
    touch ${prefix}__bvec_sh_shells

    if [[ ${task.ext.out_indices} ]]; then
        touch ${meta.id}__out_indices
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(uv pip -q -n list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """
}
