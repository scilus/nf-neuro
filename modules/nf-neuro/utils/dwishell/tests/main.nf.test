// nf-core modules test utils/dwishell
nextflow_process {

    name "Test Process UTILS_DWISHELL"
    script "../main.nf"
    process "UTILS_DWISHELL"

    tag "modules"
    tag "modules_nfneuro"
    tag "utils"
    tag "utils/dwishell"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "DWIms1000-2000-dir10-10.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("dwishell - extract single shell") {

        config "./nextflow_single_shell.config"

        when {
            process {
            """
                input[0] = LOAD_DATA.out.test_data_directory
                    .map{ test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/dwi.nii.gz"),
                        file("\${test_data_directory}/dwi.bval"),
                        file("\${test_data_directory}/dwi.bvec")
                    ]}
            """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.dwi,
                    process.out.bval,
                    process.out.bvec,
                    process.out.versions).match() }
            )
        }
    }

    test("dwishell - extract two shells") {

        config "./nextflow_two_shells.config"

        when {
            process {
            """
                input[0] = LOAD_DATA.out.test_data_directory
                    .map{ test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/dwi.nii.gz"),
                        file("\${test_data_directory}/dwi.bval"),
                        file("\${test_data_directory}/dwi.bvec")
                    ]}
            """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.dwi,
                    process.out.bval,
                    process.out.bvec,
                    process.out.versions).match() }
            )
        }
    }

    test("dwishell - extract out indices") {

        config "./nextflow_out_indices.config"

        when {
            process {
            """
                input[0] = LOAD_DATA.out.test_data_directory
                    .map{ test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/dwi.nii.gz"),
                        file("\${test_data_directory}/dwi.bval"),
                        file("\${test_data_directory}/dwi.bvec")
                    ]}
            """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.dwi,
                    process.out.bval,
                    process.out.bvec,
                    process.out.out_indices,
                    process.out.versions).match() }
            )
        }
    }

    test("dwishell - stub") {

        tag "stub"
        options "-stub-run"

        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory
                    .map{ test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/dwi.nii.gz"),
                        file("\${test_data_directory}/dwi.bval"),
                        file("\${test_data_directory}/dwi.bvec")
                    ]}
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.versions).match() }
            )
        }

    }
}
