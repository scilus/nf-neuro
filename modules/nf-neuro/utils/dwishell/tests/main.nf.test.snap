{
    "dwishell - extract two shells": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__dwi_sh_shells.nii.gz:md5,021c3fdc96d8ab0311efa0e3e17e8309"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__bval_sh_shells:md5,4068ac4e3f4893fd9f63fe7be83229a0"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__bvec_sh_shells:md5,d4ddf35e80a7d2caa8e77505a729ceb8"
                ]
            ],
            [
                "versions.yml:md5,e0f2705d122d83d3a6625951bcbfd464"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-20T14:57:05.985127779"
    },
    "dwishell - extract single shell": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__dwi_sh_shells.nii.gz:md5,e714c303160c86434672e57a3ef2c0c7"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__bval_sh_shells:md5,ad6aa09708a6772442779d6f80b0a2a1"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__bvec_sh_shells:md5,d2253329eeb22ab5f60f360f6bb40a72"
                ]
            ],
            [
                "versions.yml:md5,e0f2705d122d83d3a6625951bcbfd464"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-20T14:57:00.024365812"
    },
    "dwishell - stub": {
        "content": [
            [
                "versions.yml:md5,e0f2705d122d83d3a6625951bcbfd464"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-20T14:57:18.230279221"
    },
    "dwishell - extract out indices": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__dwi_sh_shells.nii.gz:md5,664892dc7a55f15dc9f92be1762fa3f8"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__bval_sh_shells:md5,fd3db13552e55a8ee6c7e8f11d7698eb"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__bvec_sh_shells:md5,2c1de45ddc899932e289a8eeb397380a"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__out_indices:md5,b2fcd56b5ae04dce90945c030520ae3f"
                ]
            ],
            [
                "versions.yml:md5,e0f2705d122d83d3a6625951bcbfd464"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-20T14:57:12.134213782"
    }
}